src/inmanta/server/agentmanager.py

```diff
@@ -213,108 +179,92 @@ async def _unpause_agent(self, env: data.Environment, endpoint: Optional[str] =
                     if session:
                         self.tid_endpoint_to_session[key] = session
                         await session.get_client().set_state(agent_name, enabled=True)
 
-    # Agent Management
-    async def ensure_agent_registered(self, env: data.Environment, nodename: str) -> data.Agent:
-        """
-            Make sure that an agent has been created in the database
-        """
-        async with self.session_lock:
-            agent = await data.Agent.get(env.id, nodename)
-            if agent is not None:
-                return agent
-            else:
-                return await self._create_default_agent(env, nodename)
+    # Notify from session listener
+    async def new_session(self, session: protocol.Session) -> None:
+        await self._register_session(session, datetime.now())
 
-    async def _create_default_agent(self, env: data.Environment, nodename: str) -> data.Agent:
-        """
-            This method creates a new agent (agent in the model) in the database.
-            If an active agent instance exists for the given agent, it is marked as a the
-            primary instance for that agent in the database.
+    # Notify from session listener
+    def expire(self, session: protocol.Session, timeout: float) -> None:
+        self.add_background_task(self._expire_session(session, datetime.now()))
 
-            Note: This method must be called under session lock
-        """
-        saved = data.Agent(environment=env.id, name=nodename, paused=False)
-        await saved.insert()
+    # Notify from session listener
+    def seen(self, session: protocol.Session) -> None:
+        self.add_background_task(self._seen_session(session))
 
-        key = (env.id, nodename)
-        session = self.tid_endpoint_to_session.get(key, None)
-        if session is not None:
-            await self._log_primary_to_db(env, [(nodename, session)], datetime.now())
+    # Seen
+    async def _seen_session(self, session: protocol.Session) -> None:
+        endpoints_with_new_primary: List[Tuple[str, Optional[uuid.UUID]]] = []
+        async with self.session_lock:
+            endpoints_in_agent_manager = self.endpoints_for_sid[session.id]
+            endpoints_in_session = set(session.endpoint_names)
+            endpoints_to_add = list(endpoints_in_session - endpoints_in_agent_manager)
+            LOGGER.debug("Adding endpoints %s to session %s on %s", endpoints_to_add, session.id, session.nodename)
+            endpoints_to_remove = list(endpoints_in_agent_manager - endpoints_in_session)
+            LOGGER.debug("Removing endpoints %s from session %s on %s", endpoints_to_add, session.id, session.nodename)
+
+            endpoints_with_new_primary += await self._failover_endpoints(session, endpoints_to_remove)
+            endpoints_with_new_primary += await self._ensure_primary_if_not_exists(session)
+            self.endpoints_for_sid[session.id] = set(session.endpoint_names)
```

Comment:
```suggestion
            self.endpoints_for_sid[session.id] = set(endpoints_in_session)
```

session.endpoint_names can be changed outside the lock